File: src/utils/course.js
/**
 * 课程模块 API
 */

import { request } from './cloud-db.js'

const COURSES_COLLECTION = 'courses'

/**
 * 获取课程列表
 */
export async function getCourses(params = {}) {
  try {
    const res = await request('GET', '/api/courses', params)
    const list = res?.list || []
    return {
      ok: true,
      list,
      total: list.length
    }
  } catch (error) {
    console.error('获取课程失败:', error)
    return { ok: false, message: error.message, list: [], total: 0 }
  }
}

/**
 * 获取课程分类
 */
export async function getCategories() {
  try {
    const res = await request('GET', '/api/courses/categories')
    return {
      ok: true,
      list: res?.list || []
    }
  } catch (error) {
    console.error('获取分类失败:', error)
    return { ok: false, list: [] }
  }
}

/**
 * 获取课程详情
 */
export async function getCourseDetail(id) {
  try {
    const res = await request('GET', `/api/courses/${id}`)
    if (res?.data) {
      return { ok: true, data: res.data }
    }
    return { ok: false, message: '课程不存在' }
  } catch (error) {
    console.error('获取课程详情失败:', error)
    return { ok: false, message: error.message }
  }
}

/**
 * 创建课程
 */
export async function createCourse(data) {
  try {
    const res = await request('POST', '/api/courses', data)
    return { ok: true, id: res.id }
  } catch (error) {
    return { ok: false, message: error.message }
  }
}

/**
 * 更新课程
 */
export async function updateCourse(id, data) {
  try {
    await request('PUT', `/api/courses/${id}`, data)
    return { ok: true }
  } catch (error) {
    return { ok: false, message: error.message }
  }
}

/**
 * 删除课程
 */
export async function deleteCourse(id) {
  try {
    await request('DELETE', `/api/courses/${id}`)
    return { ok: true }
  } catch (error) {
    return { ok: false, message: error.message }
  }
}

/**
 * 获取管理员添加的所有课程
 */
export async function getAdminCourses() {
  try {
    const res = await getCourses()
    if (Array.isArray(res?.list)) {
      return res.list.map(item => ({
        ...item,
        id: item.id
      }))
    }
    return []
  } catch (error) {
    console.error('获取课程失败:', error)
    return []
  }
}

/**
 * 减少课程库存
 */
export async function decreaseStock(items) {
  try {
    for (const item of items) {
      const courseId = item.courseId || item.productId || item.id
      if (!courseId) continue

      const res = await request('POST', '/api/courses/decrease-stock', [{
        courseId,
        quantity: item.quantity || 1
      }])

      if (!res?.ok) {
        console.warn('减少库存失败:', courseId)
      }
    }
    return { ok: true }
  } catch (error) {
    console.error('减少库存失败:', error)
    return { ok: false, message: error.message }
  }
}

/**
 * 增加课程库存
 */
export async function increaseStock(items) {
  try {
    for (const item of items) {
      const courseId = item.courseId || item.productId || item.id
      if (!courseId) continue

      const res = await request('POST', '/api/courses/increase-stock', [{
        courseId,
        quantity: item.quantity || 1
      }])

      if (!res?.ok) {
        console.warn('增加库存失败:', courseId)
      }
    }
    return { ok: true }
  } catch (error) {
    console.error('增加库存失败:', error)
    return { ok: false, message: error.message }
  }
}

/**
 * 购买课程
 */
export async function purchaseCourse(courseId, userId) {
  try {
    const res = await request('POST', '/api/course-orders', {
      course_id: courseId,
      user_id: userId
    })
    if (res?.ok) {
      return { ok: true, data: res.data }
    }
    return { ok: false, message: res?.detail || '购买失败' }
  } catch (error) {
    console.error('购买课程失败:', error)
    return { ok: false, message: error.message }
  }
}

/**
 * 检查课程购买状态
 */
export async function checkCoursePurchased(userId, courseId) {
  try {
    const res = await request('GET', `/api/course-orders/check/${courseId}`)
    if (res?.ok) {
      return {
        ok: true,
        purchased: res.purchased,
        course_link: res.course_link,
        course_type: res.course_type
      }
    }
    return { ok: false, purchased: false }
  } catch (error) {
    console.error('检查购买状态失败:', error)
    return { ok: false, purchased: false }
  }
}

/**
 * 购买内部课程（带视频）
 */
export async function purchaseInternalCourse(courseId) {
  try {
    const res = await request('POST', '/api/course-orders/internal', {
      course_id: courseId
    })
    if (res?.ok) {
      return { ok: true, data: res.data }
    }
    return { ok: false, message: res?.detail || '购买失败' }
  } catch (error) {
    console.error('购买内部课程失败:', error)
    return { ok: false, message: error.message }
  }
}

/**
 * 检查内部课程购买状态
 */
export async function checkInternalCoursePurchased(courseId) {
  try {
    const res = await request('GET', `/api/course-orders/check-internal/${courseId}`)
    if (res?.ok) {
      return {
        ok: true,
        purchased: res.purchased,
        video_url: res.video_url
      }
    }
    return { ok: false, purchased: false }
  } catch (error) {
    console.error('检查内部课程购买状态失败:', error)
    return { ok: false, purchased: false }
  }
}

/**
 * 获取用户已购买的课程列表
 */
export async function getPurchasedCourses(userId) {
  try {
    const res = await request('GET', '/api/course-orders/purchased')
    if (res?.ok) {
      return { ok: true, list: res.list || [] }
    }
    return { ok: false, list: [] }
  } catch (error) {
    console.error('获取已购课程失败:', error)
    return { ok: false, list: [] }
  }
}

/**
 * 获取用户的课程订单
 */
export async function getCourseOrders(userId) {
  try {
    const res = await request('GET', '/api/course-orders/user')
    if (res?.ok) {
      return { ok: true, list: res.list || [] }
    }
    return { ok: false, list: [] }
  } catch (error) {
    console.error('获取课程订单失败:', error)
    return { ok: false, list: [] }
  }
}

File: backend/app/routers/course.py
"""
课程路由
"""
import time
from typing import Optional, List
from fastapi import APIRouter, HTTPException, Query, Body

from app.models.course import Course, Category
from app.models.cart import Cart
from app.schemas.schemas import CourseCreate, CourseUpdate

router = APIRouter(prefix="/api/courses", tags=["课程"])


def _c(course):
    """转换课程数据格式"""
    return {
        "id": course.id,
        "title": course.title,
        "desc": course.desc,
        "cover": course.cover,
        "images": course.images,
        "price": str(course.price),
        "priceNow": str(course.price_now),
        "price_now": str(course.price_now),
        "unit": course.unit,
        "stock": course.stock,
        "category": course.category,
        "tags": course.tags,
        "instructor": course.instructor,
        "rating": str(course.rating) if course.rating else None,
        "studentCount": course.student_count,
        "student_count": course.student_count,
        "duration": course.duration,
        "level": course.level,
        "video_url": course.video_url,
        "videoUrl": course.video_url,
        "status": course.status,
        "sales": course.sales,
        "is_featured": course.is_featured,
        "is_hot": course.is_hot,
    }


# ========== 课程分类接口 ==========

@router.get("/categories", response_model=dict)
async def get_categories():
    """获取所有课程分类"""
    categories = await Category.all().filter(status="on").order_by("sort")
    return {
        "ok": True,
        "list": [{"id": c.id, "name": c.name, "icon": c.icon, "color": c.color} for c in categories]
    }


@router.post("/categories", response_model=dict)
async def create_category(name: str, icon: str = None, color: str = None, sort: int = 0):
    """创建课程分类"""
    now = int(time.time() * 1000)
    cat = await Category.create(name=name, icon=icon, color=color, sort=sort, status="on", created_at=now)
    return {"ok": True, "id": cat.id}


# ========== 课程接口 ==========

# 获取课程列表
@router.get("")
async def get_courses(
    category: str = Query(None, description="分类名称"),
    keyword: str = Query(None, description="搜索关键词"),
    status: str = Query("on", description="状态筛选"),
):
    """获取课程列表，支持分类筛选和关键词搜索"""
    query = Course.all()

    if status:
        query = query.filter(status=status)

    if category:
        query = query.filter(category=category)

    if keyword:
        query = query.filter(title__icontains=keyword)

    courses = await query.order_by("-created_at")
    return {
        "ok": True,
        "list": [_c(c) for c in courses],
        "total": len(courses)
    }


# 获取课程详情
@router.get("/{course_id}")
async def get_course(course_id: int):
    course = await Course.filter(id=course_id).first()
    if not course:
        raise HTTPException(status_code=404, detail="课程不存在")

    return {
        "ok": True,
        "data": _c(course)
    }


# 创建课程
@router.post("")
async def create_course(data: CourseCreate):
    now = int(time.time() * 1000)
    course = await Course.create(
        title=data.title,
        desc=data.desc,
        cover=data.cover,
        images=data.images,
        price=data.price,
        price_now=data.price_now,
        unit=data.unit or "课程",
        stock=data.stock or 0,
        category=data.category,
        tags=data.tags,
        instructor=data.instructor,
        rating=data.rating,
        duration=data.duration,
        level=data.level,
        video_url=data.video_url,
        status="on",
        created_at=now,
        updated_at=now
    )

    return {"ok": True, "id": course.id}


# 更新课程
@router.put("/{course_id}")
async def update_course(course_id: int, data: CourseUpdate):
    course = await Course.filter(id=course_id).first()
    if not course:
        raise HTTPException(status_code=404, detail="课程不存在")

    update_data = {k: v for k, v in data.model_dump().items() if v is not None}
    update_data["updated_at"] = int(time.time() * 1000)

    await course.update_from_dict(update_data)
    await course.save()

    return {"ok": True}


# 删除课程
@router.delete("/{course_id}")
async def delete_course(course_id: int):
    course = await Course.filter(id=course_id).first()
    if not course:
        raise HTTPException(status_code=404, detail="课程不存在")

    await course.delete()
    return {"ok": True}


# ========== 库存管理 ==========

@router.post("/decrease-stock")
async def decrease_stock(items: list):
    """减少课程库存"""
    for item in items:
        course_id = item.get("courseId") or item.get("productId") or item.get("id")
        quantity = item.get("quantity", 1)

        course = await Course.filter(id=course_id).first()
        if course:
            new_stock = max(0, course.stock - quantity)
            await Course.filter(id=course_id).update(stock=new_stock)

    return {"ok": True}


@router.post("/increase-stock")
async def increase_stock(items: list):
    """增加课程库存"""
    for item in items:
        course_id = item.get("courseId") or item.get("productId") or item.get("id")
        quantity = item.get("quantity", 1)

        course = await Course.filter(id=course_id).first()
        if course:
            new_stock = course.stock + quantity
            await Course.filter(id=course_id).update(stock=new_stock)

    return {"ok": True}


@router.get("/{course_id}/stock")
async def get_course_stock(course_id: int):
    course = await Course.filter(id=course_id).first()
    if not course:
        return {"ok": False, "message": "课程不存在"}

    return {"ok": True, "stock": course.stock}


# ========== 精选和热门课程 ==========

@router.get("/featured/list")
async def get_featured_courses():
    """获取精选课程列表"""
    courses = await Course.filter(status="on", is_featured=True).order_by("-updated_at")
    return {
        "ok": True,
        "list": [_c(c) for c in courses],
        "total": len(courses)
    }


@router.get("/hot/list")
async def get_hot_courses():
    """获取热门课程列表"""
    courses = await Course.filter(status="on", is_hot=True).order_by("-updated_at")
    return {
        "ok": True,
        "list": [_c(c) for c in courses],
        "total": len(courses)
    }


@router.post("/set-featured")
async def set_featured_courses(body: dict = Body(...)):
    """设置精选课程（替换模式）"""
    # 支持 { course_ids: [...] } 格式
    course_ids = body.get("course_ids", body) if isinstance(body, dict) else body
    if not isinstance(course_ids, list):
        course_ids = []
    # 先取消所有精选
    await Course.all().update(is_featured=False)
    # 再设置新的精选
    if course_ids:
        await Course.filter(id__in=course_ids).update(is_featured=True)
    return {"ok": True, "message": "精选课程已更新"}


@router.post("/set-hot")
async def set_hot_courses(body: dict = Body(...)):
    """设置热门课程（替换模式）"""
    # 支持 { course_ids: [...] } 格式
    course_ids = body.get("course_ids", body) if isinstance(body, dict) else body
    if not isinstance(course_ids, list):
        course_ids = []
    # 先取消所有热门
    await Course.all().update(is_hot=False)
    # 再设置新的热门
    if course_ids:
        await Course.filter(id__in=course_ids).update(is_hot=True)
    return {"ok": True, "message": "热门课程已更新"}
